### PR TITLE
Fix bug for f_new update condition of pq_astar

### LINKED ISSUE #13: question about pq_astar
I was looking at this particular [line](https://github.com/omron-sinicx/neural-astar/blob/6d88e7ca4c00447862da4650e5b339489371e3c4/src/neural_astar/planner/pq_astar.py#L136) and wondering, if there is a bug in this code. Shouldn't we also consider the condition where a node is already in the open list, and maybe has lesser fnew than the one in the open list ?

### PR DESCRIPTION
**Pull request by yonetaniryo**

### CODE REVIEW

File: src/neural_astar/planner/pq_astar.py [hunk 1]
```diff
@@ -133,19 +133,28 @@ def solve_single(
...
                 )
-                open_list.additem(n_idx, fnew)
-                parent_list[n_idx] = v_idx
+
+                # conditions for the nodes not yet in the open list nor closed list
+                cond = (n_idx not in open_list) & (n_idx not in close_list)
+
+                # condition for the nodes already in the open list but with larger f value
+                if n_idx in open_list:
+                    cond = cond | (open_list[n_idx] > f_new)
```
Comment (yonetaniryo) [inline src/neural_astar/planner/pq_astar.py]: This condition was missed in the initial implementation, as suggested in #14

**Comment by HareshKarnan** [comment]
Thanks for this PR! could you explain the f_new equation ? I'm finding it hard to understand intuitively.

**Comment by yonetaniryo** [comment]
> Thanks for this PR! could you explain the f_new equation ? I'm finding it hard to understand intuitively.

Sure. Let me explain in the following way.

First of all, `open_list` in [L121](https://github.com/omron-sinicx/neural-astar/blob/fda7e63d9d21031c7985263f22e098cf20a3bcc1/src/neural_astar/planner/pq_astar.py#L121) is a priority queue that stores the f value of opened nodes. In [L133](https://github.com/omron-sinicx/neural-astar/blob/fda7e63d9d21031c7985263f22e098cf20a3bcc1/src/neural_astar/planner/pq_astar.py#L133), `v_cost` and `v_idx` are the f value of the selected node and its index, respectively. 

You know, `f = g + h`, and let me describe `f_selected = g_selected + h_selected` for the selected node and `f_neighbor = g_neighbor + h_neighbor` for the neighbor nodes of the selected node, where `f_selected = v_cost` and `f_neighbor = f_new` in this implementation. Also, `g_neighbor = g_selected + c_neighbor`, where `c_neighbor` is a node cost imposed when arriving at the neighbor node and `c_neighbor = pred_cost_vct[n_idx]`. 

Now, we can see that `f_neighbor = g_selected + c_neighbor + h_neighbor = f_selected - h_selected + c_neighbor + h_neighbor`, which is implemented in the following lines. https://github.com/omron-sinicx/neural-astar/blob/fda7e63d9d21031c7985263f22e098cf20a3bcc1/src/neural_astar/planner/pq_astar.py#L138-L143

Maybe renaming `v_cost` -> `f_selected` and `v_idx` -> `idx_selected` will make it a bit more intuitive?

**Review by HareshKarnan — APPROVED** [review]
Looks good to me!

**Comment by yonetaniryo** [comment]
Thank you @HareshKarnan for your review and support! Please feel free to raise issues or proposals if you find any.

**Pull request merged**